active/003-seq-feature.md

```diff
@@ -0,0 +1,201 @@
+        interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+        interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+        interval_metadata.update(Feature(gene='sagB', location=0),
+                                 Feature(gene='sagB', location=1))
+```
+
+
+`add(feature, *intervals)`
+- `feature` : `skbio.Feature`. new feature object add into the `IntervalMetadata` object
+- `*intervals` : an iterable of tuples corresponding to intervals where the feature object exists.
+- This allows for a single feature (include those that have non-continugous features) to be added into the `IntervalMetadata` object.
+
+```python
+   interval_metadata = IntervalMetadata()
+   interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+   interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+```
+
+`query(*args, **kwargs)`
+- `*args` : an iterable of tuples to search for features
+- `**kwargs` : keyword arguments to query features by their attributes.
+- This allows for features to be searched by both intervals and keywords (i.e. CDS vs exon, ...)
+- Note that for a specific interval query, multiple features can be returned.
+ 
+```python
+   interval_metadata = IntervalMetadata(features={
+                                        Feature(gene='sagA', location=0): [(0, 2), (4, 7)],
+                                        Feature(gene='sagB', location=0): [(3, 5)]})
+ 
+   feats = interval_metadata.query((1, 2))
+   feats = interval_metadata.query(gene='sagB')
+```
+
+This is the current implementation in micronota and fits all of our use-cases at the moment.  We don't feel that IntervalIndex objects as suggested in the recent pandas poll request can easily handle all of our use cases (i.e. non-contiguous intervals isn't straightforward).
+
+##Drawbacks
+- The `query` method does a linear lookup when searching for attributes over features.  This problem can blow up (in terms of development) if we want faster querying.  Which is why we opted for slow, linear search.  Again suggestions here are welcome.
+- features and intervals are loosely coupled.
+
+#Unresolved Questions (arranged in priority)
+- The `update` method may not be the most straightforward for users.  The reason why we opted for this design was because the `skbio.Feature` object can only be indexed uniquely by hashing the entire feature.  Suggestions on how to improve this will be welcome.
+- Is there a better way to associate intervals to their corresponding features?
+- How will features be deleted?
+- How will intervals for a given feature be updated?
+- How exactly will slicing be handled?  May want to be able to pass intervals/features into `__getitem__` within the `skbio.Sequence` class
+- How exactly should features be added to the interval_metadata?
+- What IntervalTree data structure should we stick with? 
+- Should we re-invent the wheel and roll out our own IntervalTree object.
+
+
+# Alternative design
```

Comment:
agree. will clean it up if we agree on this design.

On Wed, Apr 27, 2016 at 2:47 PM, Evan Bolyen <notifications@github.com>
wrote:

> In active/003-seq-feature.md
> <https://github.com/biocore/scikit-bio-rfcs/pull/9#discussion_r61342446>:
>
> > +##Drawbacks
> > +- The `query` method does a linear lookup when searching for attributes over features.  This problem can blow up (in terms of development) if we want faster querying.  Which is why we opted for slow, linear search.  Again suggestions here are welcome.
> > +- features and intervals are loosely coupled.
> > +
> > +#Unresolved Questions (arranged in priority)
> > +- The `update` method may not be the most straightforward for users.  The reason why we opted for this design was because the `skbio.Feature` object can only be indexed uniquely by hashing the entire feature.  Suggestions on how to improve this will be welcome.
> > +- Is there a better way to associate intervals to their corresponding features?
> > +- How will features be deleted?
> > +- How will intervals for a given feature be updated?
> > +- How exactly will slicing be handled?  May want to be able to pass intervals/features into `__getitem__` within the `skbio.Sequence` class
> > +- How exactly should features be added to the interval_metadata?
> > +- What IntervalTree data structure should we stick with?
> > +- Should we re-invent the wheel and roll out our own IntervalTree object.
> > +
> > +
> > +# Alternative design
>
> If this is the preferred design, I would flip the order in this document.
> (I think we are converging on something really nice)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio-rfcs/pull/9/files/aa684efdb3e4a27fc5babf73190f048e738f3f18#r61342446>
>